### ISSUE TITLE
ESBMC-CHERI failed to build

### ISSUE THREAD

**Issue by kunjsong01** [esbmc/esbmc#1609]
ESBMC-CHERI failed to build due to parsing error when generating libc model. 

### Error log:
Please see file attached below: 
[error_log.txt](https://github.com/esbmc/esbmc/files/14000716/error_log.txt)

### CHERI build command used: 
```
python cheribuild.py cheribsd-riscv64-purecap disk-image-riscv64-purecap -d
```

### CMake configuration command used: 

Note that the combo 'cheribsd-riscv64-purecap disk-image-riscv64-purecap' doesn't build any rootfs*-hybrid
 in the output directory. So I used "/home/kunjian/cheri/output/rootfs-riscv64-purecap" for both `-DESBMC_CHERI_HYBRID_SYSROOT` and `-DESBMC_CHERI_PURECAP_SYSROOT` as shown below: 

```
cd esbmc && mkdir build && cd build && cmake .. -GNinja -DBUILD_TESTING=On -DENABLE_REGRESSION=On $ESBMC_CLANG -DBUILD_STATIC=OFF -DBoolector_DIR=$PWD/../../boolector-release -DZ3_DIR=$PWD/../../z3 -DENABLE_MATHSAT=ON -DMathsat_DIR=$PWD/../../mathsat -DENABLE_YICES=On -DYices_DIR=$PWD/../../yices -DCVC4_DIR=$PWD/../../cvc4 -DGMP_DIR=$PWD/../../gmp -DENABLE_BITWUZLA=OFF -DCMAKE_INSTALL_PREFIX:PATH=$PWD/../../release \
-DESBMC_CHERI_HYBRID_SYSROOT=/home/kunjian/cheri/output/rootfs-riscv64-purecap \
-DESBMC_CHERI_PURECAP_SYSROOT=/home/kunjian/cheri/output/rootfs-riscv64-purecap
```

I also tried to build ESBMC-CHERI with just `-DESBMC_CHERI_PURECAP_SYSROOT` solely for purecap CHERI rootfs, but it also failed with parsing error, please see log attached below: 

[error_log_without_hybrid.txt](https://github.com/esbmc/esbmc/files/14000814/error_log_without_hybrid.txt)

**Comment by lucasccordeiro** [comment]
Hi @fbrausse,

Can I ask you whether you would have any insights here?

**Comment by lucasccordeiro** [comment]
@fbrausse: If I follow the instructions at https://github.com/esbmc/esbmc/blob/master/BUILDING.md, I get errors after executing this command:

````
cmake -GNinja -S ../llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_PARALLEL_LINK_JOBS=4 -DLLVM_CCACHE_BUILD=FALSE -DLLVM_INSTALL_BINUTILS_SYMLINKS=TRUE -DLLVM_ENABLE_LIBXML2=FALSE -DLLVM_ENABLE_ZLIB=FORCE_ON -DLLVM_ENABLE_OCAMLDOC=FALSE -DLLVM_ENABLE_BINDINGS=FALSE -DLLVM_INCLUDE_EXAMPLES=FALSE -DLLVM_INCLUDE_DOCS=FALSE -DLLVM_INCLUDE_BENCHMARKS=FALSE -DCLANG_ENABLE_STATIC_ANALYZER=FALSE -DCLANG_ENABLE_ARCMT=FALSE -DLLVM_ENABLE_Z3_SOLVER=FALSE -DLLVM_TOOL_LLVM_MCA_BUILD=FALSE -DLLVM_TOOL_LLVM_EXEGESIS_BUILD=FALSE -DLLVM_TOOL_LLVM_RC_BUILD=FALSE -DLLVM_ENABLE_LLD=TRUE -DLLVM_OPTIMIZED_TABLEGEN=FALSE -DLLVM_USE_SPLIT_DWARF=TRUE -DLLVM_ENABLE_ASSERTIONS=TRUE '-DLLVM_LIT_ARGS=--max-time 3600 --timeout 300 -s -vv' '-DLLVM_TARGETS_TO_BUILD=AArch64;ARM;Mips;RISCV;X86;host' -DENABLE_EXPERIMENTAL_NEW_PASS_MANAGER=FALSE -DCLANG_ROUND_TRIP_CC1_ARGS=FALSE '-DLLVM_ENABLE_PROJECTS=llvm;clang;lld' -DCMAKE_INSTALL_PREFIX=../../clang13 -DCMAKE_C_COMPILER=/usr/bin/cc -DCMAKE_CXX_COMPILER=/usr/bin/c++ -DCMAKE_ASM_COMPILER=/usr/bin/cc -DCMAKE_BUILD_RPATH_USE_ORIGIN=TRUE
````

````
-- Performing Test CXX_SUPPORTS_CUSTOM_LINKER - Failed
CMake Error at cmake/modules/HandleLLVMOptions.cmake:291 (message):
  Host compiler does not support '-fuse-ld=lld'
Call Stack (most recent call first):
  CMakeLists.txt:701 (include)


-- Configuring incomplete, errors occurred!
````

Can I ask you whether you would have any recommendations here?

**Comment by fbrausse** [comment]
For reference, for cheribuild locally I've been using the `llvm-project` and `cheribsd` on tag `cheri-rel-20210817` and the `morello-llvm-project` on commit aba2e0847d0dbc33d7292876c2c22c2d610d5359. I needed to apply this patch in order to get cheribsd itself to build: [cheribsd-20210817-werror.patch.txt](https://github.com/esbmc/esbmc/files/14035400/cheribsd-20210817-werror.patch.txt)

Will try a more recent version of the upstream CTSRD tools.

**Comment by lucasccordeiro** [comment]
@fbrausse: Can I ask you whether you could update us here with the ESBMC-CHERI build? What has been done? What did work? What didn't work?

**Comment by XLiZHI** [comment]
> @fbrausse: If I follow the instructions at https://github.com/esbmc/esbmc/blob/master/BUILDING.md, I get errors after executing this command:
> 
> ```
> cmake -GNinja -S ../llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_PARALLEL_LINK_JOBS=4 -DLLVM_CCACHE_BUILD=FALSE -DLLVM_INSTALL_BINUTILS_SYMLINKS=TRUE -DLLVM_ENABLE_LIBXML2=FALSE -DLLVM_ENABLE_ZLIB=FORCE_ON -DLLVM_ENABLE_OCAMLDOC=FALSE -DLLVM_ENABLE_BINDINGS=FALSE -DLLVM_INCLUDE_EXAMPLES=FALSE -DLLVM_INCLUDE_DOCS=FALSE -DLLVM_INCLUDE_BENCHMARKS=FALSE -DCLANG_ENABLE_STATIC_ANALYZER=FALSE -DCLANG_ENABLE_ARCMT=FALSE -DLLVM_ENABLE_Z3_SOLVER=FALSE -DLLVM_TOOL_LLVM_MCA_BUILD=FALSE -DLLVM_TOOL_LLVM_EXEGESIS_BUILD=FALSE -DLLVM_TOOL_LLVM_RC_BUILD=FALSE -DLLVM_ENABLE_LLD=TRUE -DLLVM_OPTIMIZED_TABLEGEN=FALSE -DLLVM_USE_SPLIT_DWARF=TRUE -DLLVM_ENABLE_ASSERTIONS=TRUE '-DLLVM_LIT_ARGS=--max-time 3600 --timeout 300 -s -vv' '-DLLVM_TARGETS_TO_BUILD=AArch64;ARM;Mips;RISCV;X86;host' -DENABLE_EXPERIMENTAL_NEW_PASS_MANAGER=FALSE -DCLANG_ROUND_TRIP_CC1_ARGS=FALSE '-DLLVM_ENABLE_PROJECTS=llvm;clang;lld' -DCMAKE_INSTALL_PREFIX=../../clang13 -DCMAKE_C_COMPILER=/usr/bin/cc -DCMAKE_CXX_COMPILER=/usr/bin/c++ -DCMAKE_ASM_COMPILER=/usr/bin/cc -DCMAKE_BUILD_RPATH_USE_ORIGIN=TRUE
> ```
> 
> ```
> -- Performing Test CXX_SUPPORTS_CUSTOM_LINKER - Failed
> CMake Error at cmake/modules/HandleLLVMOptions.cmake:291 (message):
>   Host compiler does not support '-fuse-ld=lld'
> Call Stack (most recent call first):
>   CMakeLists.txt:701 (include)
> 
> 
> -- Configuring incomplete, errors occurred!
> ```
> 
> Can I ask you whether you would have any recommendations here?

Hi Dr @lucasccordeiro,

May I ask if you still have this problem?
 
This may be because `lld` is not installed on the system. I tried the following command to solve the problem:

```
sudo apt-get install lld
```

**Comment by lucasccordeiro** [comment]
@XLiZHI,

I can try to build it again.

Would you like me to follow the instructions in https://github.com/esbmc/esbmc/blob/master/BUILDING.md?

Shall we add this command `sudo apt-get install lld` to the instructions?

**Comment by XLiZHI** [comment]
> Would you like me to follow the instructions in https://github.com/esbmc/esbmc/blob/master/BUILDING.md?

Yes, I think the instructions are mostly correct.

Before run:
```
python 3 cheribuild.py cheribsd-riscv64-purecap disk-image-riscv64-purecapit-d
```
In addition to the [setup](https://github.com/CTSRD-CHERI/cheribuild/blob/main/README.md) it provides, It's better check to see if the `bc` tool is installed:

```
sudo apt-get install bc
```
Cheri's file size is very large, close to 30GB, which takes a lot of time.

Currently I build with commend as follows:

```
cmake .. -GNinja -DBUILD_TESTING=On -DENABLE_REGRESSION=On $ESBMC_CLANG -DBUILD_STATIC=OFF -DBoolector_DIR=$PWD/../../boolector-release -DENABLE_MATHSAT=Off -DENABLE_YICES=Off -DCVC4_DIR=$PWD/../../cvc4 -DENABLE_BITWUZLA=OFF -DCMAKE_INSTALL_PREFIX:PATH=$PWD/../../release -DESBMC_CHERI_PURECAP_SYSROOT=$HOME/cheri/output/rootfs-riscv64-purecap -DESBMC_CHERI=On -DCMAKE_BUILD_TYPE=Sanitizer -DSANITIZER_TYPE=UBSAN
```

I had an error when building it earlier, so I added this flag (`-DCMAKE_BUILD_TYPE=Sanitizer -DSANITIZER_TYPE=UBSAN`), but it won't build the debug version (I think this is a temporary solution):
```
/home/lxzy/ESBMC_Project/esbmc/build/src/catch2/internal/catch_test_case_info_hasher.cpp:19: undefined reference to `__ubsan_handle_type_mismatch_v1'
```

I'm not sure whether the Franz patch in order to solve this problem https://github.com/esbmc/esbmc/issues/1609#issuecomment-1907626067.

> Shall we add this command sudo apt-get install lld to the instructions?

I think we can update the instructions together after building successfully.

**Comment by fbrausse** [comment]
With the recent fix in #1974, may I ask whether this build problem still exists?

**Comment by XLiZHI** [comment]
The build issue has been fixed, thank you @fbrausse
